sw/device/tests/crypto/cryptotest/BUILD

```diff
@@ -0,0 +1,122 @@
+#     for key_len in ("128", "192", "256")
+# ]
+
+# testvector_args = " ".join([
+#     "--aes-json=\"$(rootpath {})\"".format(target)
+#     for target in testvector_targets
+# ])
+
+# opentitan_test(
+#     name = "aes_kat_test",
+#     cw310 = cw310_params(
+#         timeout = "long",
+#         binaries = {"//sw/device/tests/crypto/cryptotest/firmware:firmware": "firmware"},
+#         data = testvector_targets,
+#         test_cmd = """
+#             --bootstrap={firmware}
+#             --bitstream={bitstream}
+#         """ + testvector_args,
+#         test_harness = "//sw/host/tests/crypto/aes_nist_kat:harness",
+#     ),
+#     exec_env = {
+#         "//hw/top_earlgrey:fpga_cw310_test_rom": None,
+#     },
+# )
+
+KAT_TYPE = [
+    "varkey",
+    "gfsbox",
+    "vartxt",
+    "keysbox",
+]
+
+KEY_LEN = [
+    "128",
+    "192",
+    "256",
+]
+
+TESTVECTOR_TARGETS = {
+    "cbc": [
+        "//sw/host/cryptotest/testvectors/data/aes_nist_kat:cbc_{}_{}_json".format(kat_type, key_len)
+        for kat_type in KAT_TYPE
+        for key_len in KEY_LEN
+    ],
+    "cfb128": [
+        "//sw/host/cryptotest/testvectors/data/aes_nist_kat:cfb128_{}_{}_json".format(kat_type, key_len)
+        for kat_type in KAT_TYPE
+        for key_len in KEY_LEN
+    ],
+    "ecb": [
+        "//sw/host/cryptotest/testvectors/data/aes_nist_kat:ecb_{}_{}_json".format(kat_type, key_len)
+        for kat_type in KAT_TYPE
+        for key_len in KEY_LEN
+    ],
+    "ofb": [
+        "//sw/host/cryptotest/testvectors/data/aes_nist_kat:ofb_{}_{}_json".format(kat_type, key_len)
+        for kat_type in KAT_TYPE
+        for key_len in KEY_LEN
+    ],
+}
+
+TESTVECTOR_ARGS = {
+    "cbc": " ".join([
+        "--aes-json=\"$(rootpath {})\"".format(target)
+        for target in TESTVECTOR_TARGETS["cbc"]
+    ]),
+    "cfb128": " ".join([
+        "--aes-json=\"$(rootpath {})\"".format(target)
+        for target in TESTVECTOR_TARGETS["cfb128"]
+    ]),
+    "ecb": " ".join([
+        "--aes-json=\"$(rootpath {})\"".format(target)
+        for target in TESTVECTOR_TARGETS["ecb"]
+    ]),
+    "ofb": " ".join([
+        "--aes-json=\"$(rootpath {})\"".format(target)
+        for target in TESTVECTOR_TARGETS["ofb"]
+    ]),
+}
+
+[
+    opentitan_test(
+        name = "aes_{}_kat".format(mode),
+        cw310 = cw310_params(
+            timeout = "long",
+            binaries = {"//sw/device/tests/crypto/cryptotest/firmware:firmware": "firmware"},
+            data = TESTVECTOR_TARGETS[mode],
+            test_cmd = """
+                --bootstrap={firmware}
+                --bitstream={bitstream}
+            """ + TESTVECTOR_ARGS[mode],
+            test_harness = "//sw/host/tests/crypto/aes_nist_kat:harness",
+        ),
+        exec_env = {
+            "//hw/top_earlgrey:fpga_cw310_test_rom": None,
+        },
+    )
+    for mode in [
+        "cbc",
+        "cfb128",
+        "ecb",
+        "ofb",
+    ]
```

Comment:
I noticed that in a few places CTR mode is missing; is there a reason to exclude it or test it separately?

Reply:
No particular reason. It's just that the batch of testvectors I got from NIST didn't include CTR tests, so cryptotest supports passing it in as an argument, but I don't currently have testvectors for it.

Reply:
I noticed the same and indeed there are no test vectors provided as part of NIST CAVP. Instead the [PDF document](https://csrc.nist.rip/CSRC/media/Projects/Cryptographic-Algorithm-Validation-Program/documents/aes/AESAVS.pdf) says:
> **Appendix A. Counter Mode Requirements**
It is a requirement of the Counter Mode that a unique counter block be used for each
plaintext block that is ever encrypted with a given key, across all messages. Due to the
fact that Counter Mode does not specify the counter that is used, it is not possible to
implement an automated test for this mode. Also, since the tests in this document are
only a statistical sampling, additional assurances are needed that the counter blocks used
are in fact unique. Therefore, if the module being tested derives counter values
internally, the testing lab shall conduct a design review of the internal counter mechanism
used to ensure that it provides unique counter block values as well as a code review of the
implementation of the counter mechanism to ensure that it properly implements the
design. In addition, the tests for ECB Mode shall be performed on the encryption engine
used for the counter mode.

Our AES hardware block implements the Standard Incrementing Function discussed in the [Recommendation for Block Cipher Modes of Operation](https://nvlpubs.nist.gov/nistpubs/legacy/sp/nistspecialpublication800-38a.pdf):
> **Appendix B: Generation of Counter Blocks**
...
**B.1 The Standard Incrementing Function**
The standard incrementing function can apply either to an entire block or to a part of a block.
Let m be the number of bits in the specific part of the block to be incremented; thus, m is a
positive integer such that m ≤ b. Any string of m bits can be regarded as the binary representation
of a non-negative integer x that is strictly less than 2^m. The standard incrementing function takes
[x]_m and returns [x+1 mod 2^m]_m .

This document further provides some test vectors, see Appendix F.5 CTR Example Vectors on Page 55. I recommend to include those vectors if somehow possible.

Reply:
Ah thanks for the detailed pointers!

I am on board with including the example vectors that you linked from NIST 800-38A, but I might leave that to a different PR since it would be a different source of testvectors and I would just grab all the other examples too while we're at it. I can make a note to return to this.

In the section you quoted from the NIST CAVP PDF, I have a question specifically about "the tests for ECB Mode shall be performed on the encryption engine used for the counter mode." Does this mean I should just run the ECB testvectors through the AES block with CTR mode? If so, I can make that addition in this PR.

Reply:
You can build CTR out of ECB and an incrementing function; it looks roughly like:
```
for i=0..nblocks:
    ciphertext[i] = plaintext[i] ^ ECB_encrypt(ctr)
    ctr = (ctr + 1) mod 2^128
```
So it wouldn't quite work to run the ECB testvectors, since the value that goes through ECB is the counter, not the plaintext. Adding the other test vectors in a separate PR sounds good though :slightly_smiling_face:

Reply:
Yes, I agree with @jadephilipoom and I must admit that I don't understand what exactly they meant in the NIST CAVP PDF. Adding the other vectors in a different PR sounds good.

Sorry for my late reply @milesdai , I was out sick the entire week.